src/bsp/src/esp/src/SpiStm.cpp

```diff
@@ -0,0 +1,198 @@
+#include "SpiStm.h"
+#include "c-common/software_crc.h"
+#include "esp32/rom/crc.h"
+#include "hal/pin_map.h"
+#include "hal/spi_callbacks.h"
+#include <cstring>
+#include <driver/gpio.h>
+
+/** These macros are used to convert the units of the size of a buffer from a number of words to a
+ * number of bytes and conversely from a size in bytes to a number of words, and similarly between a
+ * sizes in bits and bytes.
+ */
+#define WORDS_TO_BYTES(words) ((words << 2U))
+#define BYTES_TO_WORDS(bytes) ((bytes >> 2U))
+#define BYTES_TO_BITS(bytes) ((bytes << 3U))
+
+void task(void* context) {
+    while (true) {
+        static_cast<SpiStm*>(context)->execute();
+    }
+}
+
+SpiStm::SpiStm(ILogger& logger) :
+    m_logger(logger), m_driverTask("stm_spi_driver", tskIDLE_PRIORITY + 1, task, this) {
+    m_txState = transmitState::SENDING_HEADER;
+    m_rxState = receiveState::RECEIVING_HEADER;
+    m_inboundMessage = {0};
+    m_outboundMessage = {0};
+    m_inboundHeader = {0};
+    m_outboundHeader = {0};
+    m_isBusy = false;
+
+    setCallback(SpiStm::transactionCallback, this);
+
+    m_driverTask.start();
+}
+
+bool SpiStm::send(const uint8_t* buffer, uint16_t length) {
+
+    if (isBusy()) {
+        return false; // Not available
+    } else if (length >= STM_SPI_MAX_MESSAGE_LENGTH) { // Message too long
+        m_logger.log(LogLevel::Error,
+                     "StmEsp: Message length of %d is larger than maximum allowed of %d", length,
+                     STM_SPI_MAX_MESSAGE_LENGTH);
+        return false;
+    } else {
+        m_logger.log(LogLevel::Debug, "Sending message of length %d", length);
+        // Memcpy necessary to have buffer word-alligned for transfer
+        std::memcpy(m_outboundMessage.m_data.data(), buffer, length);
```

Comment:
Could you just link the doc here, like this link
https://docs.espressif.com/projects/esp-idf/en/latest/esp32/api-reference/peripherals/spi_slave.html#restrictions-and-known-issues